### ISSUE TITLE
http3  - ResponseController/SetWriteDeadline timeout not detected client side

### ISSUE THREAD

**Issue by kgersen** [quic-go/quic-go#3892]
v0.35.0 added ResponseController support.
We can set `SetWriteDeadline  `on a server but it times out the client just see normal termination.

poc:

        git clone https://github.com/kgersen/h3ctx
        cd h3ctx
        go run main.go -st 1s

This will do a test with HTTP/2 then HTTP/3. 
In both case, after 1 second the server timeout (`SetWriteDeadline`). 
With quic-go the client doesn't know (reading the request body end normally with err is nil). Server side we do get an error.
With `net/http` we get an error client side too.

If haven't looked at what the official 'semantic' and behavior should be. 

But it's handy to know client side if a deadline occurred on the server. For instance if you download a file you don't know if you got it all or not. You need to use other meta  information like size or checksum which aren't always available and requires extra steps.

**Comment by kgersen** [comment]
tl;dr version:
 - `net/http` with a `SetWriteDeadline(timeout)` : client "http get" a big file -> timeout server side -> error client side
 - same case with `quic-go` : client "http/3 get" a big file-> timeout server side -> no error client side (we just get the beginning of the file up to the timeout)

**Comment by Glonee** [comment]
See https://datatracker.ietf.org/doc/html/rfc9114#name-request-cancellation-and-re
Looks like we didn't use any error code at `SetWriteDeadline` timeout.

**Comment by francislavoie** [comment]
FYI @dunglas
